### PR TITLE
Multiple changes for shadowsocks-android

### PR DESCRIPTION
**Pull request by madeye**
This PR added:

1. Protect socket logic for Android.
2. Misc fixes for Android VPNService.

I think some rework and cleanup are still needed, especially for the interfaces.

### CODE REVIEW

File: src/relay/sys/unix/mod.rs [hunk 1]
```diff
@@ -25,8 +34,82 @@ pub fn sockaddr_to_std(saddr: &libc::sockaddr_storage) -> io::Result<SocketAddr>
...
+
+    Ok(socket)
+}
+
 /// Create a `UdpSocket` binded to `addr`
 #[inline(always)]
 pub async fn create_udp_socket(addr: &SocketAddr) -> io::Result<UdpSocket> {
-    UdpSocket::bind(addr).await
+    let socket = UdpSocket::bind(addr).await?;
+    Ok(socket)
```
Comment (zonyitoo) [inline src/relay/sys/unix/mod.rs]: Hmm? Why is that?

**Comment by zonyitoo** [comment]
Windows build is broken. :(

File: src/relay/sys/unix/mod.rs (hunk 1 above)
Comment (madeye) [reply src/relay/sys/unix/mod.rs]: Yeah, we should revert the change here. I did some experiment here and forgot to revert...

**Comment by madeye** [comment]
I fixed the build on Windows. But it looks the test still failed. Not sure why it happens.

**Comment by zonyitoo** [comment]
`net2` is deprecated, use `socket2` instead.

Like this: https://github.com/shadowsocks/shadowsocks-rust/blob/master/src/relay/tcprelay/sys/unix/linux.rs#L82

**Comment by zonyitoo** [comment]
Excellent. I am going to do some code cleanup later.

**Pull request merged**